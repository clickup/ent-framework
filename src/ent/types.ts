import type { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  ID,
  LoadByInput,
  Order,
  Table,
  UniqueKey,
  Where,
} from "../types";
import type { Validation } from "./Validation";
import type { VC } from "./VC";

/**
 * A very shallow interface of Ent class (as a collection of static methods).
 * Used in some places where we need the very minimum from the Ent.
 */
export interface EntClass<TTable extends Table = any> {
  readonly SCHEMA: Schema<TTable>;
  readonly VALIDATION: Validation<TTable>;
  readonly name: string; // class constructor name

  new (): Ent;
  loadX(vc: VC, id: string): Promise<Ent>;
  loadNullable(vc: VC, id: string): Promise<Ent | null>;
  loadIfReadableNullable(vc: VC, id: string): Promise<Ent | null>;
  count(vc: VC, where: CountInput<TTable>): Promise<number>;
  exists(vc: VC, where: ExistsInput<TTable>): Promise<boolean>;
  select(
    vc: VC,
    where: Where<TTable>,
    limit: number,
    order?: Order<TTable>
  ): Promise<Ent[]>;
  selectChunked(
    vc: VC,
    where: Where<TTable>,
    chunkSize: number,
    limit: number,
    custom?: {}
  ): AsyncIterableIterator<Ent[]>;
  loadByX(vc: VC, keys: LoadByInput<TTable, UniqueKey<TTable>>): Promise<Ent>;
}

/**
 * A very shallow interface of one Ent.
 */
export interface Ent {
  readonly [ID]: string;
  readonly vc: VC;
  deleteOriginal(): Promise<boolean>;
}
