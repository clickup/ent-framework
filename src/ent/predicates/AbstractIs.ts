import type { Predicate } from "./Predicate";

/**
 * A base interface for all user validation predicates.
 */
export interface AbstractIs<TRow> extends Predicate<TRow> {
  /** The field this validation predicate is related to (null means that it
   * applies to the entire Ent). */
  readonly field: string | null;
  /** In case the predicate returns false or doesn't provide error messages by
   * throwing EntValidationError, this message will be used. When message is
   * null, it means that we expect the validator to return detailed information
   * about each field errored (e.g. ValidatorStandardSchemaResult).  */
  readonly message: string | null;
}

/**
 * Result of plain validators.
 */
export type ValidatorPlainResult = boolean;

/**
 * Result of Zod safeParse() calls.
 */
export type ValidatorZodSafeParseResult =
  | {
      readonly success: true;
      readonly error?: undefined;
    }
  | {
      readonly success: false;
      readonly error: {
        readonly issues: ReadonlyArray<{
          readonly message: string;
          readonly path: readonly PropertyKey[];
        }>;
      };
    };

/**
 * Result of a Standard Schema validators: https://standardschema.dev/
 */
export type ValidatorStandardSchemaResult =
  | {
      readonly value: unknown;
      readonly issues?: undefined;
    }
  | {
      readonly issues: Array<{
        readonly message: string;
        readonly path?:
          | ReadonlyArray<PropertyKey | { readonly key: PropertyKey }>
          | undefined;
      }>;
    };
