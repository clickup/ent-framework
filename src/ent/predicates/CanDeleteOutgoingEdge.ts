import type { EntClass } from "../types";
import type { VC } from "../VC";
import type { Predicate } from "./Predicate";
import { IDsCacheDeletable } from "./Predicate";

/**
 * Checks that an Ent available via a field can be deleted, or Ent doesn't exist
 * (e.g. Ent is orphaned). See CanReadOutgoingEdge comments for more details.
 */
export class CanDeleteOutgoingEdge<TField extends string>
  implements Predicate<Record<TField, string | null>>
{
  readonly name;

  constructor(
    public readonly field: TField,
    public readonly toEntClass: EntClass,
  ) {
    this.name = this.constructor.name + "(" + this.field + ")";
  }

  async check(vc: VC, row: Record<TField, string | null>): Promise<boolean> {
    const toID = row[this.field];
    if (!toID) {
      return false;
    }

    const cache = vc.cache(IDsCacheDeletable);
    if (cache.has(toID!)) {
      return true;
    }

    // Load the target Ent and check that it's deletable.
    const toEnt = await this.toEntClass.loadNullable(vc, toID!);
    if (toEnt === null) {
      return true;
    }

    await this.toEntClass.VALIDATION.validateDelete(vc, toEnt);

    // Sill here and not thrown? save to the cache.
    cache.add(toID!);
    return true;
  }
}
