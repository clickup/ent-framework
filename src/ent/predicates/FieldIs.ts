import type { VC } from "../VC";
import type {
  AbstractIs,
  ValidatorPlainResult,
  ValidatorStandardSchemaResult,
  ValidatorZodSafeParseResult,
} from "./AbstractIs";
import { maybeThrowEntValidationError } from "./internal/maybeThrowEntValidationError";

/**
 * A field validator function that returns a boolean.
 */
export type FieldIsValidatorPlain<
  TField extends string,
  TRow extends Partial<Record<TField, unknown>>,
> = (
  fieldValue: TRow[TField],
  row: TRow,
  vc: VC,
) => ValidatorPlainResult | Promise<ValidatorPlainResult>;

/**
 * A field validator function that returns a Zod result.
 */
export type FieldIsValidatorZodSafeParse<TRow> = (
  fieldValue: unknown,
  row: TRow,
  vc: VC,
) => ValidatorZodSafeParseResult | Promise<ValidatorZodSafeParseResult>;

/**
 * A field validator function that returns a Standard Schema validation result.
 */
export type FieldIsValidatorStandardSchemaV1<TRow> = (
  fieldValue: unknown,
  row: TRow,
  vc: VC,
) => ValidatorStandardSchemaResult | Promise<ValidatorStandardSchemaResult>;

/**
 * Checks that the validator function returns true for the value in some field.
 */
export class FieldIs<
  TField extends string,
  TRow extends Partial<Record<TField, unknown>>,
> implements AbstractIs<TRow>
{
  readonly name;
  readonly field: TField;
  readonly message: string | null;
  readonly validator;

  /**
   * Manual validator. Implies that we can trust the fieldValue TS type.
   */
  constructor(
    field: TField,
    validator: FieldIsValidatorPlain<TField, TRow>,
    message: string,
  );

  /**
   * Rich validator, like Standard Schema (https://standardschema.dev) or Zod.
   * No implications are made on the fieldValue type.
   */
  constructor(
    field: TField,
    validator:
      | FieldIsValidatorStandardSchemaV1<TRow>
      | FieldIsValidatorZodSafeParse<TRow>,
  );

  constructor(
    field: TField,
    validator:
      | FieldIsValidatorPlain<TField, TRow>
      | FieldIsValidatorZodSafeParse<TRow>
      | FieldIsValidatorStandardSchemaV1<TRow>,
    message?: string,
  ) {
    this.name = this.constructor.name + "(" + field + ")";
    this.field = field;
    this.message = message ?? null;
    this.validator = validator;
  }

  /**
   * Returns true if validation succeeds. Returns false if it wants the client
   * to use this.message as a validation failure response. Throws an instance of
   * EntValidationError when it needs to deliver the detailed error messages
   * about multiple fields.
   */
  async check(vc: VC, row: TRow): Promise<boolean> {
    const res = await this.validator(row[this.field], row, vc);
    return maybeThrowEntValidationError({
      name: this.name,
      field: this.field,
      res,
      allowRichResult: this.message === null,
    });
  }
}
