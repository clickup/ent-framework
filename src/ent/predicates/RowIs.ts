import type { VC } from "../VC";
import type {
  AbstractIs,
  ValidatorPlainResult,
  ValidatorStandardSchemaResult,
  ValidatorZodSafeParseResult,
} from "./AbstractIs";
import { maybeThrowEntValidationError } from "./internal/maybeThrowEntValidationError";

/**
 * A row validator function that returns a boolean.
 */
export type RowIsValidatorPlain<TRow> = (
  row: TRow,
  vc: VC,
) => ValidatorPlainResult | Promise<ValidatorPlainResult>;

/**
 * A row validator function that returns a Zod result.
 */
export type RowIsValidatorZodSafeParse<TRow> = (
  row: TRow,
  vc: VC,
) => ValidatorZodSafeParseResult | Promise<ValidatorZodSafeParseResult>;

/**
 * A row validator function that returns a Standard Schema validation result.
 */
export type RowIsValidatorStandardSchemaV1<TRow> = (
  row: TRow,
  vc: VC,
) => ValidatorStandardSchemaResult | Promise<ValidatorStandardSchemaResult>;

/**
 * Checks that the validator function returns true for the entire row.
 */
export class RowIs<TRow> implements AbstractIs<TRow> {
  readonly name;
  readonly field = null;
  readonly message: string | null;
  readonly validator;

  /**
   * Manual validator.
   */
  constructor(validator: RowIsValidatorPlain<TRow>, message: string);

  /**
   * Rich validator, like Standard Schema (https://standardschema.dev) or Zod.
   */
  constructor(
    validator:
      | RowIsValidatorZodSafeParse<TRow>
      | RowIsValidatorStandardSchemaV1<TRow>,
  );

  constructor(
    validator:
      | RowIsValidatorPlain<TRow>
      | RowIsValidatorZodSafeParse<TRow>
      | RowIsValidatorStandardSchemaV1<TRow>,
    message?: string,
  ) {
    this.name = this.constructor.name;
    this.message = message ?? null;
    this.validator = validator;
  }

  /**
   * Returns true if validation succeeds. Returns false if it wants the client
   * to use this.message as a validation failure response. Throws an instance of
   * EntValidationError when it needs to deliver the detailed error messages
   * about multiple fields.
   */
  async check(vc: VC, row: TRow): Promise<boolean> {
    const res = await this.validator(row, vc);
    return maybeThrowEntValidationError({
      name: this.name,
      field: null,
      res,
      allowRichResult: this.message === null,
    });
  }
}
