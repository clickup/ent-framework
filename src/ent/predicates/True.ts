import { VC } from "../VC";
import { Predicate } from "./Predicate";

/**
 * Always passes; used for e.g. globally accessed objects.
 */
export class True implements Predicate<never> {
  readonly name = this.constructor.name;

  async check(_vc: VC, _row: object): Promise<boolean> {
    return true;
  }
}
