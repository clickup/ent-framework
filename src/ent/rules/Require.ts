import type { VC } from "../VC";
import type { RuleResult } from "./Rule";
import { Rule, RuleDecision } from "./Rule";

/**
 * Returns TOLERATE if the predicate succeeds, otherwise DENY.
 * - Used mostly for write permission checks.
 * - This rule may still throw an exception if it's a wild one (i.e. not derived
 *   from EntAccessError).
 */
export class Require<TInput extends object> extends Rule<TInput> {
  readonly _TAG!: "Require";

  async evaluate(vc: VC, input: TInput): Promise<RuleResult> {
    return (await this.predicate.check(vc, input))
      ? { decision: RuleDecision.TOLERATE, rule: this, cause: null }
      : { decision: RuleDecision.DENY, rule: this, cause: null };
  }
}
