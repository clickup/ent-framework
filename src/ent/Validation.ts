import { indent } from "../helpers";
import {
  InsertFieldsRequired,
  InsertInput,
  Row,
  Table,
  UpdateInput,
} from "../types";
import { EntNotInsertableError } from "./errors/EntNotInsertableError";
import { EntNotReadableError } from "./errors/EntNotReadableError";
import { EntNotUpdatableError } from "./errors/EntNotUpdatableError";
import {
  EntValidationError,
  EntValidationErrorInfo,
} from "./errors/EntValidationError";
import { Predicate } from "./predicates/Predicate";
import { Require } from "./rules/Require";
import { evaluate, Rule, RuleDecision, RuleResult } from "./rules/Rule";
import { buildNewRow } from "./Triggers";
import { VC } from "./VC";

export type ValidationRules<TTable extends Table> = {
  readonly tenantUserIDField?: InsertFieldsRequired<TTable> & string;
  readonly load: Validation<TTable>["load"];
  readonly insert: Validation<TTable>["insert"];
  readonly update?: Validation<TTable>["update"];
  readonly delete?: Validation<TTable>["delete"];
  readonly validate?: Array<Predicate<Row<TTable>> & EntValidationErrorInfo>;
};

export class Validation<TTable extends Table> {
  readonly tenantUserIDField?: InsertFieldsRequired<TTable> & string;
  readonly load: Array<Rule<Row<TTable>>>;
  readonly insert: Array<Rule<InsertInput<TTable>>>;
  readonly update: Array<Rule<Row<TTable>>>;
  readonly delete: Array<Rule<Row<TTable>>>;
  readonly validate: Array<Require<Row<TTable>>>;

  constructor(private entName: string, rules: ValidationRules<TTable>) {
    this.load = rules.load;
    this.insert = rules.insert;
    this.update = rules.update || (this.insert as any);
    this.delete = rules.delete || this.update;
    this.validate = (rules.validate || []).map((pred) => new Require(pred));
    this.tenantUserIDField = rules.tenantUserIDField;
  }

  async validateLoad(vc: VC, row: Row<TTable>): Promise<void> {
    await this.validatePrivacyImpl(
      "load",
      this.load,
      vc,
      row,
      false,
      EntNotReadableError
    );
  }

  async validateInsert(vc: VC, input: InsertInput<TTable>): Promise<void> {
    await this.validateUserInputImpl(vc, input as Row<TTable>);
    await this.validatePrivacyImpl(
      "insert",
      this.insert,
      vc,
      input,
      true, // parallel
      EntNotInsertableError
    );
  }

  async validateUpdate(
    vc: VC,
    old: Row<TTable>,
    input: UpdateInput<TTable>,
    privacyOnly = false
  ): Promise<void> {
    // Simulate the update, as if it's applied to the ent.
    const newRow = buildNewRow(old, input);

    if (!privacyOnly) {
      await this.validateUserInputImpl(vc, newRow);
    }

    await this.validatePrivacyImpl(
      "update",
      this.update,
      vc,
      newRow,
      true, // parallel
      EntNotUpdatableError
    );
  }

  async validateDelete(vc: VC, row: Row<TTable>): Promise<void> {
    await this.validatePrivacyImpl(
      "delete",
      this.delete,
      vc,
      row,
      true, // parallel
      EntNotUpdatableError // same exception as for update
    );
  }

  private async validatePrivacyImpl(
    op: string,
    rules: Array<Rule<object>>,
    vc: VC,
    row: object,
    parallel: boolean,
    ExceptionClass:
      | typeof EntNotReadableError
      | typeof EntNotInsertableError
      | typeof EntNotUpdatableError
  ) {
    this.validateTenantUserIDImpl(vc, row, ExceptionClass);

    const { allow, results } = await evaluate(rules, vc, row, parallel);
    if (allow) {
      return;
    }

    throw new ExceptionClass(this.entName, vc.toString(), row as any, {
      message: resultsToText(op, results),
    });
  }

  private validateTenantUserIDImpl(
    vc: VC,
    row: object,
    ExceptionClass:
      | typeof EntNotReadableError
      | typeof EntNotInsertableError
      | typeof EntNotUpdatableError
  ) {
    if (this.tenantUserIDField === undefined) {
      return;
    }

    const rowTenantUserID = (row as any)[this.tenantUserIDField];
    if (rowTenantUserID === vc.userID) {
      return;
    }

    throw new ExceptionClass(this.entName, vc.toString(), row as any, {
      message:
        this.tenantUserIDField +
        " is expected to be " +
        JSON.stringify(vc.userID) +
        ", but got " +
        JSON.stringify(rowTenantUserID),
    });
  }

  private async validateUserInputImpl(vc: VC, row: Row<TTable>) {
    const { allow, results } = await evaluate(this.validate, vc, row, true);
    if (allow) {
      return;
    }

    const failedPredicates = results
      .filter((result) => result.decision === RuleDecision.DENY)
      .map((result) => result.rule.predicate as any as EntValidationErrorInfo);
    if (failedPredicates.length > 0) {
      throw new EntValidationError(this.entName, failedPredicates);
    }
  }
}

/**
 * A helper function which returns a debugging text for a list of rule
 * evaluation results.
 */
function resultsToText(op: string, results: RuleResult[]) {
  if (results.length === 0) {
    return `No "${op}" rules defined`;
  }

  return results
    .map(
      ({ rule, decision, cause }) =>
        "Rule " +
        rule.name +
        " returned " +
        decision +
        (cause ? ", because:\n" + indent(cause.message) : "")
    )
    .join("\n");
}
