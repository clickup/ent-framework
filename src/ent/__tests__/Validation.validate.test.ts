import { EntNotFoundError, EntValidationError } from "..";
import { PgSchema } from "../../pg/PgSchema";
import { ID } from "../../types";
import type {
  ValidatorStandardSchemaResult,
  ValidatorZodSafeParseResult,
} from "../predicates/AbstractIs";
import { FieldIs } from "../predicates/FieldIs";
import { RowIs } from "../predicates/RowIs";
import { True } from "../predicates/True";
import { Require } from "../rules/Require";
import type { ValidationRules } from "../Validation";
import { Validation } from "../Validation";
import { ValidationTester } from "./test-utils";

const companySchema = new PgSchema(
  'ent.validation"validate"company',
  {
    id: { type: ID, autoInsert: "gen_id()" },
    tenant_id: { type: ID },
    name: { type: String, autoInsert: "''" },
  },
  [],
);

let tester: ValidationTester;

beforeEach(() => {
  tester = new ValidationTester();
});

test("0010: update fails when field validation fails", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new FieldIs(
          "tenant_id",
          (value, row) =>
            tester.respond("FieldIs(tenant_id)", false, [value, row]),
          "some one",
        ),
        new FieldIs(
          "tenant_id",
          (value, row) =>
            tester.respond("FieldIs(tenant_id)", false, [value, row]),
          "some two",
        ),
        new FieldIs("name", async (value, row) =>
          tester.respond(
            "FieldIs(tenant_id)",
            standardSchemaIssuesAsync("name", "boom"),
            [value, row],
          ),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateUpdate",
      updateInput: { tenant_id: "42", name: "hi" },
    }),
  ).toBe(false);
});

test("0020: update succeeds when validation fails, but the field is untouched", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new FieldIs(
          "name",
          (value, row) => tester.respond("FieldIs(name)", false, [value, row]),
          "boom",
        ),
        new FieldIs("name", async (value, row) =>
          tester.respond(
            "FieldIs(name)",
            standardSchemaIssues("name", "boom"),
            [value, row],
          ),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateUpdate",
      updateInput: { tenant_id: "101" },
    }),
  ).toBe(true);
});

test("0030: insert succeeds when multiple validators succeed", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs(
          (row) => tester.respond("RowIs", true, row),
          "always succeeds",
        ),
        new RowIs((row) => tester.respond("RowIs", zodSuccess(), row)),
        new FieldIs(
          "name",
          (value, row) => tester.respond("FieldIs(name)", true, [value, row]),
          "always succeeds",
        ),
        new FieldIs("name", async (value, row) =>
          tester.respond("FieldIs(name)", standardSchemaSuccess(), [
            value,
            row,
          ]),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(true);
});

test("0040: insert fails when field validation fails", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new FieldIs(
          "name",
          (value, row) => tester.respond("FieldIs(name)", false, [value, row]),
          "boom",
        ),
        new FieldIs("name", async (value, row) =>
          tester.respond("FieldIs(name)", zodIssues("name", "boom"), [
            value,
            row,
          ]),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

test("0050: update fails when RowIs fails", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs((row) => tester.respond("RowIs", false, row), "boom"),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateUpdate",
      updateInput: {},
    }),
  ).toBe(false);
});

test("0060: update fails when RowIs fails, and RowIs propagates field name", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs((row) =>
          tester.respond("RowIs", zodIssues("name", "boom"), row),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateUpdate",
      updateInput: { name: "hi" },
    }),
  ).toBe(false);
});

test("0070: update succeeds when RowIs fails, but the field it propagates is untouched", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs((row) =>
          tester.respond("RowIs", standardSchemaIssues("name", "boom"), row),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateUpdate",
      updateInput: {},
    }),
  ).toBe(true);
});

test("0080: insert fails when RowIs fails", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs((row) => tester.respond("RowIs", false, row), "boom"),
        new RowIs((row) =>
          tester.respond("RowIs", zodIssues("name", "boom"), row),
        ),
        new RowIs((row) =>
          tester.respond(
            "RowIs",
            zodIssues(null, "Expected object, received number"),
            row,
          ),
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

test("0090: insert throws when FieldIs throws a weird error", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new FieldIs(
          "name",
          (value, row) =>
            tester.respond("FieldIs", Error("ouch"), [value, row]),
          "boom",
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

test("0095: insert fails on EntNotFoundError in FieldIs", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new FieldIs(
          "name",
          (value, row) =>
            tester.respond(
              "FieldIs",
              new EntNotFoundError("MyEnt", { id: "42" }),
              [value, row],
            ),
          "boom",
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

test("0096: insert fails on manual EntValidationError in RowIs", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs(
          (row) =>
            tester.respond(
              "RowIs",
              new EntValidationError("MyEnt", [
                { field: "name", message: "huh" },
              ]),
              row,
            ),
          "boom",
        ),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

test("0100: insert throws when RowIs throws a weird error", async () => {
  expect(
    await tester.matchSnapshot({
      validation: createValidation([
        new RowIs((row) => tester.respond("RowIs", Error("ouch"), row), "boom"),
      ]),
      row: { id: "123", tenant_id: "42", name: "hi" },
      method: "validateInsert",
    }),
  ).toBe(false);
});

function createValidation(
  validate: ValidationRules<typeof companySchema.table>["validate"],
): Validation<typeof companySchema.table> {
  return new Validation<typeof companySchema.table>("table", {
    inferPrincipal: async (vc) => vc.toGuest(),
    load: [],
    insert: [new Require(new True())],
    validate,
  });
}

function zodSuccess(): ValidatorZodSafeParseResult {
  return { success: true };
}

function zodIssues(
  field: string | null,
  message: string,
): ValidatorZodSafeParseResult {
  return {
    success: false,
    error: { issues: [{ message, path: field ? [field] : [] }] },
  };
}

function standardSchemaSuccess(): ValidatorStandardSchemaResult {
  return { value: 42 };
}

function standardSchemaIssues(
  field: string,
  message: string,
): ValidatorStandardSchemaResult {
  return { issues: [{ message, path: [field] }] };
}

async function standardSchemaIssuesAsync(
  field: string,
  message: string,
): Promise<ValidatorStandardSchemaResult> {
  return standardSchemaIssues(field, message);
}
