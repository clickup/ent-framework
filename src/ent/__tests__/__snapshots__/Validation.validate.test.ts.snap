// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0010: update fails when field validation fails 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: tenant_id: "some one", tenant_id: "some two", name: "boom"
--- (what actually happened) ---
[FieldIs(tenant_id) [ '42', { id: '123', tenant_id: '42', name: 'hi' } ] returned] false
[FieldIs(tenant_id) [ '42', { id: '123', tenant_id: '42', name: 'hi' } ] returned] false
[FieldIs(tenant_id) [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] returned] 'Promise'"
`;

exports[`0020: update succeeds when validation fails, but the field is untouched 1`] = `
"OK
--- (what actually happened) ---
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '101', name: 'hi' } ] returned] false
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '101', name: 'hi' } ] returned] { issues: [ { message: 'boom', path: [ 'name' ] } ] }"
`;

exports[`0030: insert succeeds when multiple validators succeed 1`] = `
"OK
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] true
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] { success: true }
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] returned] true
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] returned] { value: 42 }"
`;

exports[`0040: insert fails when field validation fails 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: name: "boom", name: "boom"
--- (what actually happened) ---
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] returned] false
[FieldIs(name) [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] returned] { success: false, error: { issues: [ { message: 'boom', path: [ 'name' ] } ] } }"
`;

exports[`0050: update fails when RowIs fails 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: null: "boom"
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] false"
`;

exports[`0060: update fails when RowIs fails, and RowIs propagates field name 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: name: "boom"
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] { success: false, error: { issues: [ { message: 'boom', path: [ 'name' ] } ] } }"
`;

exports[`0070: update succeeds when RowIs fails, but the field it propagates is untouched 1`] = `
"OK
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] { issues: [ { message: 'boom', path: [ 'name' ] } ] }"
`;

exports[`0080: insert fails when RowIs fails 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: null: "boom", name: "boom", null: "Expected object, received number"
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] false
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] { success: false, error: { issues: [ { message: 'boom', path: [ 'name' ] } ] } }
[RowIs { id: '123', tenant_id: '42', name: 'hi' } returned] { success: false, error: { issues: [ { message: 'Expected object, received number', path: [] } ] } }"
`;

exports[`0090: insert throws when FieldIs throws a weird error 1`] = `
"Error Thrown Through
--- (error returned to client) ---
Error: ouch
--- (what actually happened) ---
[FieldIs [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] threw] Error: ouch"
`;

exports[`0095: insert fails on EntNotFoundError in FieldIs 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: name: "EntNotFoundError"
--- (what actually happened) ---
[FieldIs [ 'hi', { id: '123', tenant_id: '42', name: 'hi' } ] threw] EntNotFoundError: MyEnt not found: id=42"
`;

exports[`0096: insert fails on manual EntValidationError in RowIs 1`] = `
"Failure
--- (error returned to client) ---
EntValidationError: table: name: "huh"
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } threw] EntValidationError: MyEnt: name: "huh""
`;

exports[`0100: insert throws when RowIs throws a weird error 1`] = `
"Error Thrown Through
--- (error returned to client) ---
Error: ouch
--- (what actually happened) ---
[RowIs { id: '123', tenant_id: '42', name: 'hi' } threw] Error: ouch"
`;
